api/auth/signer/v4/chunked.go

```diff
@@ -7,52 +7,116 @@ package v4
+
+	// ErrMissingSeparator appears if chunk header doesn't contain ';' separator.
+	ErrMissingSeparator = errors.New("missing header separator")
+
+	// ErrNoChunksSeparator appears if chunks not properly separated between each other.
+	// They should be divided with \r\n bytes.
+	ErrNoChunksSeparator = errors.New("no chunk separator")
+)
 
 // NewChunkedReader returns a new chunkedReader that translates the data read from r
 // out of HTTP "chunked" format before returning it.
 // The chunkedReader returns io.EOF when the final 0-length chunk is read.
-//
-// NewChunkedReader is not needed by normal applications. The http package
-// automatically decodes chunking when reading response bodies.
-func NewChunkedReader(r io.Reader) io.Reader {
-	br, ok := r.(*bufio.Reader)
-	if !ok {
-		br = bufio.NewReader(r)
-	}
-	return &chunkedReader{r: br}
+func NewChunkedReader(r io.ReadCloser, streamSigner *StreamSigner) io.ReadCloser {
+	return &chunkedReader{
+		r: bufio.NewReader(r),
+		// bufio.Reader can't be closed, thus left link to the original reader to close it later.
+		origReader:     r,
+		streamSigner:   streamSigner,
+		chunkSignature: make([]byte, signatureHexLength),
+	}
 }
 
 type chunkedReader struct {
-	r        *bufio.Reader
-	n        uint64 // unread bytes in chunk
-	err      error
-	buf      [2]byte
-	checkEnd bool // whether need to check for \r\n chunk footer
+	chunkData      hash.Hash
+	chunkSignature []byte
+	r              *bufio.Reader
+	origReader     io.ReadCloser
+	n              uint64 // unread bytes in chunk
+	err            error
+	buf            [2]byte
+	checkEnd       bool // whether need to check for \r\n chunk footer
+	streamSigner   *StreamSigner
+}
+
+// Close implements [io.ReadCloser].
+func (cr *chunkedReader) Close() (err error) {
+	return cr.origReader.Close()
 }
 
 func (cr *chunkedReader) beginChunk() {
 	// chunk-size CRLF
-	var line []byte
-	line, cr.err = readChunkLine(cr.r)
+	var line, chunkSignature []byte
+	line, chunkSignature, cr.err = readChunkLine(cr.r)
 	if cr.err != nil {
 		return
 	}
 	cr.n, cr.err = parseHexUint(line)
 	if cr.err != nil {
 		return
 	}
+
+	if err := cr.validatePreviousChunkData(); err != nil {
+		cr.err = err
+		return
+	}
+
+	// creating instance here to avoid validating non-existent chunk in the first validatePreviousChunkData call.
+	if cr.chunkData == nil {
+		cr.chunkData = sha256.New()
+	} else {
+		cr.chunkData.Reset()
+	}
+
+	copy(cr.chunkSignature, chunkSignature)
+
 	if cr.n == 0 {
+		if err := cr.validatePreviousChunkData(); err != nil {
+			cr.err = err
+			return
+		}
+
 		cr.err = io.EOF
 	}
 }
 
+func (cr *chunkedReader) validatePreviousChunkData() error {
+	if cr.chunkData != nil {
+		calculatedSignature, err := cr.streamSigner.GetSignatureByHash(cr.chunkData)
+		if err != nil {
+			return fmt.Errorf("GetSignature: %w", err)
+		}
+
+		if !bytes.Equal(cr.chunkSignature, []byte(hex.EncodeToString(calculatedSignature))) {
```

Comment:
String comparison would be a little more efficient here, `string(cr.chunkSignature)` costs much less than `[]byte(...)`